Sources/JavaScriptKit/BasicObjects/JSPromise.swift

```diff
@@ -0,0 +1,136 @@
+    }
+
+    /// This private initializer assumes that the passed object is a JavaScript `Promise`
+    private init(unsafe object: JSObject) {
+        self.jsObject = object
+    }
+
+    /** Creates a new `JSPromise` instance from a given JavaScript `Promise` object. If `jsObject`
+    is not an instance of JavaScript `Promise`, this initializer will return `nil`.
+    */
+    public init?(_ jsObject: JSObject) {
+        guard jsObject.isInstanceOf(JSObject.global.Promise.function!) else { return nil }
+        self.jsObject = jsObject
+    }
+
+    /** Creates a new `JSPromise` instance from a given JavaScript `Promise` object. If `value`
+    is not an object and is not an instance of JavaScript `Promise`, this function will 
+    return `nil`.
+    */
+    public static func construct(from value: JSValue) -> Self? {
+        guard case let .object(jsObject) = value else { return nil }
+        return Self.init(jsObject)
+    }
+
+    /** Schedules the `success` closure to be invoked on sucessful completion of `self`.
+    */
+    public func then(success: @escaping (Success) -> ()) {
+        let closure = JSClosure {
+            success(Success.construct(from: $0[0])!)
+        }
+        callbacks.append(closure)
+        jsObject.then.function!(closure)
+    }
+
+    /** Returns a new promise created from chaining the current `self` promise with the `success`
+    closure invoked on sucessful completion of `self`. The returned promise will have a new 
+    `Success` type equal to the return type of `success`.
+    */
+    public func then<ResultType: JSValueConvertible>(
+        success: @escaping (Success) -> ResultType
+    ) -> JSPromise<ResultType, Failure> {
+        let closure = JSClosure {
+            success(Success.construct(from: $0[0])!).jsValue()
+        }
+        callbacks.append(closure)
+        return .init(unsafe: jsObject.then.function!(closure).object!)
+    }
+
+    /** Returns a new promise created from chaining the current `self` promise with the `success`
+    closure invoked on sucessful completion of `self`. The returned promise will have a new type
+    equal to the return type of `success`.
+    */
+    public func then<ResultSuccess: JSValueConvertible, ResultFailure: JSValueConstructible>(
+        success: @escaping (Success) -> JSPromise<ResultSuccess, ResultFailure>
+    ) -> JSPromise<ResultSuccess, ResultFailure> {
+        let closure = JSClosure {
+            success(Success.construct(from: $0[0])!).jsValue()
+        }
+        callbacks.append(closure)
+        return .init(unsafe: jsObject.then.function!(closure).object!)
+    }
+
+    /** Schedules the `failure` closure to be invoked on rejected completion of `self`.
+    */
+    public func `catch`(failure: @escaping (Failure) -> ()) {
+        let closure = JSClosure {
+            failure(Failure.construct(from: $0[0])!)
+        }
+        callbacks.append(closure)
+        jsObject.then.function!(JSValue.undefined, closure)
```

Comment:
Is there a reason you’re not using `jsObject.catch!(closure)` here?

Reply:
Also the difference between `foo.bar.function!(...)` and `foo.bar!(...)` is that the latter preserves the automatic `this`-binding JS does, which is important for object methods that are likely not already bound to their object.

Reply:
> Is there a reason you’re not using `jsObject.catch!(closure)` here?

[According to MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/catch):

> It behaves the same as calling `Promise.prototype.then(undefined, onRejected)` (in fact, calling `obj.catch(onRejected)` internally calls `obj.then(undefined, onRejected))`.